### ISSUE TITLE
Combining CL_DEVICE_TYPE_DEFAULT with other types in clGetDeviceIDs query

### ISSUE THREAD

**Issue by s-barannikov** [KhronosGroup/OpenCL-Docs#1060]
Hello,

Suppose the platform has two devices:

- a `CL_DEVICE_TYPE_CPU`, which is also the `CL_DEVICE_TYPE_DEFAULT` device;
- a `CL_DEVICE_TYPE_GPU` device.

Which of these two devices, if any, should be returned by `clGetDeviceIDs` query
with `device_type == CL_DEVICE_TYPE_DEFAULT | CL_DEVICE_TYPE_GPU`?

Should I return both as the documentation seems to imply? Or, is this a documentation bug and I should:
 1. return the CPU device only, reading this as _return the DEFAULT device, if one exists, and otherwise a GPU device_?
 2. return the GPU device only, reading this as _return a GPU device, if one exists, and otherwise the DEFAULT device_?
 3. return `CL_DEVICE_NOT_FOUND`, reading this as _return the DEFAULT device if it is a GPU device_?
 4. return `CL_INVALID_DEVICE_TYPE` as such query does not make sense?

IMHO such query does not make much sense because it is much likely that a user would want either the default device or a specific device, but not both. That is, `CL_DEVICE_TYPE_DEFAULT` should not be allowed to be combined with any other device types in `clGetDeviceIDs` query (and probably in `clCreateContextFromType`, too).

**Comment by SunSerega** [comment]
What about the `CL_DEVICE_TYPE_ALL` enum? It contains all the bits, including the `CL_DEVICE_TYPE_DEFAULT`.
Currently, the default context I'm creating in my wrapper uses `CL_DEVICE_TYPE_GPU` and `CL_DEVICE_TYPE_ALL` (if no GPUs are found).
My logic is that if there are multiple GPUs or no GPUs but multiple other OpenCL devices - why wouldn't I use all of them in one context? Please correct me if this logic is flawed.

**Comment by s-barannikov** [comment]
@SunSerega
I don't see a contradiction here. If CL_DEVICE_TYPE_DEFAULT is a discrete value, it should be excluded from CL_DEVICE_TYPE_ALL, too. The behavior wouldn't change because CL_DEVICE_TYPE_DEFAULT must be ORed with some other device type.

**Comment by Oblomov** [comment]
The device type is a bitfield, and the `CL_DEVICE_TYPE_*` are masks. In the given situation, you should return all devices matching _any_ of the bits, so you would return both the (CPU) default define and the GPU device. In the presented scenario, the result would the same as if the program had requested `CL_DEVICE_TYPE_CPU | CL_DEVICE_TYPE_GPU` (for which both CPU and GPU devices are returned).

**Comment by s-barannikov** [comment]
@Oblomov Thanks, I understand that. My question is whether this behavior is intended or not, because for me it looks like an oversight. Why would a user want to combine `CL_DEVICE_TYPE_DEFAULT` with anything else in `clGetDeviceIDs` query?

**Comment by bashbaug** [comment]
I don't think it's productive to debate whether this behavior was intended or not.  It's been around for way too long to change it now.  We can certainly look at clarifying the spec so the behavior is more clear, though.

For example, we could change the description of the _device_type_ parameter from this:

> * device_type is a bitfield that identifies the type of OpenCL device. The device_type can be used to query specific OpenCL devices or all OpenCL devices available. [...]

to something like this:

> * device_type specifies the types of OpenCL devices to return.  This is a bitfield mask, and OpenCL device that match any of the requested types will be returned. [...]

Does this look reasonable?  Do you have other suggestions?

Noting for completeness: I do see one bit of oddness regarding `CL_DEVICE_TYPE_ALL` and `CL_DEVICE_TYPE_CUSTOM`.  Specifically, querying `CL_DEVICE_TYPE_ALL` is is supposed to return "All OpenCL devices available in the platform, except for `CL_DEVICE_TYPE_CUSTOM` devices."

OK, but in the headers we have:

```c
#ifdef CL_VERSION_1_2
#define CL_DEVICE_TYPE_CUSTOM                       (1 << 4)
#endif
#define CL_DEVICE_TYPE_ALL                          0xFFFFFFFF
```

So, is there any way to query all devices, including custom devices?  Seems like `CL_DEVICE_TYPE_CUSTOM` should really be using an upper bit, not included in `CL_DEVICE_TYPE_ALL`, if we wanted to have the "all devices except custom devices" behavior.

**Comment by s-barannikov** [comment]
@bashbaug 
I considered it a bit more and figured that, given the ponential breaks, it is not feasible to change the current behavior. Three implementations that I've examined (Intel, AMD and POCL) handle `CL_DEVICE_TYPE_DEFAULT` in three different ways (of those, only AMD follows the spec), and such change would potentially break some of them. I'm just gonna do it the fourth way :)

> Does this look reasonable? Do you have other suggestions?

I think the current description is clear enough.

> Noting for completeness: I do see one bit of oddness regarding `CL_DEVICE_TYPE_ALL` and `CL_DEVICE_TYPE_CUSTOM`. Specifically, querying `CL_DEVICE_TYPE_ALL` is is supposed to return "All OpenCL devices available in the platform, except for `CL_DEVICE_TYPE_CUSTOM` devices."

I saw the issue (#300) and also find it odd. In fact, I'm implementing the runtime solely for devices I'd like to think of as custom (due to relaxed reqirements to memory sizes etc.). This oddity represents an issue since custom devices are not shown in e.g. `clinfo` output (one may consider this as `clinfo` bug). I had to switch to an "accelerator" type for now.
